### PR TITLE
feat: Loris configuration(M2-6644)

### PR DESCRIPTION
**Pull request by ramirlm**
- [ x ] Tests for the changes have been added
- [ x ] Delivered the `fix` or `feature` branches into `develop` or `release` branches via `Squash and Merge` (to keep clean history)

### 📝 Description

🔗 [Jira Ticket M2-6644](https://mindlogger.atlassian.net/browse/M2-6644)

This pull request includes several changes to the Loris integration feature within the Builder module. The changes encompass updates to the API, type definitions, UI components, and tests.

### API Enhancements:
* **`src/modules/Builder/api/api.ts`:** Added new API methods for fetching integration projects, saving integrations, and getting integration status. Updated existing methods to include `appletId` in the request parameters. [[1]](diffhunk://#diff-b8838ec1b9d2b323b9a2c2728d99abccad9d0f6230fd61d5725d65d09ede18f3L4-R11) [[2]](diffhunk://#diff-b8838ec1b9d2b323b9a2c2728d99abccad9d0f6230fd61d5725d65d09ede18f3L16-R24) [[3]](diffhunk://#diff-b8838ec1b9d2b323b9a2c2728d99abccad9d0f6230fd61d5725d65d09ede18f3R37-R55)
* **`src/modules/Builder/api/api.types.ts`:** Introduced new types for integration parameters and configuration, including `FetchIntegrationStatusParams`, `FetchIntegrationProjectsParams`, and `SaveIntegrationParams`. [[1]](diffhunk://#diff-0ff95ad7352a3ca26675383b562901610b8f625c159c7d7aae83891dbaa7f3edR2) [[2]](diffhunk://#diff-0ff95ad7352a3ca26675383b562901610b8f625c159c7d7aae83891dbaa7f3edR70-R98)

### UI and Component Updates:
* **`src/modules/Builder/features/BuilderAppletSettings/BuilderAppletSettings.tsx`:** Refactored the Loris integration settings to use feature flags directly and removed redundant code. [[1]](diffhunk://#diff-425c943da0dedc1dd7f53c08c99982b50ec992ce417ed1f5ecbf6fabb98f5dc2L6-R17) [[2]](diffhunk://#diff-425c943da0dedc1dd7f53c08c99982b50ec992ce417ed1f5ecbf6fabb98f5dc2L37-R30)
* **`src/modules/Builder/features/BuilderAppletSettings/BuilderAppletSettings.utils.tsx`:** Updated the settings component to use the new `LorisIntegration` component. [[1]](diffhunk://#diff-c683c73a2bfe0357939e3a383a521680fa8ed92ed0b778e7137b30e6b6c6979bL18-R18) [[2]](diffhunk://#diff-c683c73a2bfe0357939e3a383a521680fa8ed92ed0b778e7137b30e6b6c6979bL61-R63)
* **`src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationForm/ConfigurationForm.tsx`:** Added a new form component for configuring Loris integration, including fields for hostname, username, and password.
* **`src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationForm/ConfigurationForm.styles.ts`:** Added styles for the new configuration form.
* **`src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationForm/index.ts`:** Exported the new configuration form component.

### Validation and Testing:
* **`src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.schema.ts`:** Added a validation schema for the configuration form using `yup`.
* **`src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx`:** Added tests for the configuration popup to ensure proper rendering, validation, and interaction.**

src/modules/Builder/api/api.ts
Added new API methods for fetching integration status and projects, and saving integrations.
Adjusted existing API calls to include new parameters.
src/modules/Builder/api/api.types.ts
Introduced new types for integration parameters and configurations.
src/modules/Builder/features/BuilderAppletSettings/BuilderAppletSettings.tsx
Modified component logic to use new feature flags and integration settings.

### CODE REVIEW

File: src/modules/Dashboard/api/api.ts [hunk 1]
```diff
@@ -984,7 +984,7 @@ export const deleteAppletAssignmentsApi = (
   });
 
 export const enableIntegrationApi = (integrations: Integration[], signal?: AbortSignal) =>
-  authApiClient.post('/integrations/', integrations, {
+  authApiClient.post('/integrations', integrations, {
     signal,
   });
```
Comment (ChaconC) [inline src/modules/Dashboard/api/api.ts]: this endpoint doesn't exist anymore, it was replaced by `saveIntegrationToApi`

File: src/modules/Dashboard/api/api.ts [hunk 2]
```diff
@@ -994,5 +994,5 @@ export const disableIntegrationApi = (integrations: string[], signal?: AbortSign
     signal,
   };
 
-  return authApiClient.delete('/integrations/', config);
+  return authApiClient.delete('/integrations', config);
```
Comment (ChaconC) [inline src/modules/Dashboard/api/api.ts]: same for this endpoint, though the integration is not part of this PR so it's ok if you keep it for now

File: src/shared/consts.tsx [hunk 3]
```diff
@@ -437,4 +437,6 @@ export const observerStyles = {
 
 export const enum IntegrationTypes {
   Loris = 'loris',
+  LorisUpperCase = 'LORIS', // In the API the integration type is in uppercase in some endpoints and lowercase in others
+  SampleIntegration = 'sample',
```
Comment (ChaconC) [inline src/shared/consts.tsx]: What is `SampleIntegration`? If it is meant as an example please add a comment explaining usage.

File: src/modules/Builder/api/api.types.ts [hunk 4]
```diff
@@ -66,3 +67,24 @@ export type LorisUsersVisit = {
...
+};
+
+export type SaveIntegrationParams = {
+  appletId: string;
+  hostname: string;
+  project: string;
+  username: string;
+  password: string;
+  integrationType: IntegrationTypes;
+};
```
Comment (ChaconC) [inline src/modules/Builder/api/api.types.ts]: This could be refactored to better align with the BE object where `Integration` has appletId, integrationType and configuration. Where configuration is different depending on the `integrationType`.

The type for `FetchIntegrationProjectsParams` is only valid for `LorisIntegrationConfiguration`, where the `password` is optional and only used for saving the new integration.

File: src/modules/Builder/api/api.types.ts [hunk 5]
```diff
@@ -66,3 +67,24 @@ export type LorisUsersVisit = {
 export type UploadLorisUsersVisitsParams = AppletId & {
   payload: LorisUsersVisit[];
 };
+
+export type FetchIntegrationStatusParams = {
+  // currently set as snake case on the API
+  applet_id: string;
+  integration_type: IntegrationTypes;
```
Comment (ChaconC) [inline src/modules/Builder/api/api.types.ts:74]: Could you double check this please, in reviewing the backend response I see camelCase responses
CleanShot 2024-10-11 at 12 16 29@2x

File: src/shared/consts.tsx [hunk 6]
```diff
@@ -437,4 +437,6 @@ export const observerStyles = {
 
 export const enum IntegrationTypes {
   Loris = 'loris',
+  LorisUpperCase = 'LORIS', // In the API the integration type is in uppercase in some endpoints and lowercase in others
```
Comment (ChaconC) [inline src/shared/consts.tsx]: I'm not seeing where we use lowercase loris, and in any case it might be better to create a separate enum for that specific usage. I think all the new endpoints are using `LORIS`.

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/LorisIntegration.utils.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    return res.data;
+  } catch (error) {
+    console.error('Error requesting to save project', error);
+    throw new Error('Failed to save project');
+  }
+};
+
+export const hasLorisIntegrationOnState = (integrations?: Integration[]): boolean =>
+  Boolean(integrations?.some((i) => i.integrationType === IntegrationTypes.Loris));
```
Comment (ChaconC) [inline src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/LorisIntegration.utils.ts:56]: For example here the response is `LORIS` and you're comparing to `loris`
CleanShot 2024-10-11 at 12 16 29@2x

File: src/modules/Dashboard/api/api.ts (hunk 1 above)
Comment (ramirlm) [reply src/modules/Dashboard/api/api.ts]: You sure? This is the one to save it

File: src/modules/Dashboard/api/api.ts (hunk 2 above)
Comment (ramirlm) [reply src/modules/Dashboard/api/api.ts]: Makes sense, will remove it

File: src/shared/consts.tsx (hunk 3 above)
Comment (ramirlm) [reply src/shared/consts.tsx]: Just the sample as I understood you requested me to add, did I misunderstand it?

File: src/modules/Builder/api/api.types.ts (hunk 4 above)
Comment (ramirlm) [reply src/modules/Builder/api/api.types.ts]: Yeah I tried to do that, but was having issues. Will create a new interface and not use the applet one and it should be ok

File: src/modules/Builder/api/api.types.ts (hunk 5 above)
Comment (ramirlm) [reply src/modules/Builder/api/api.types.ts:74]: Only if it is in a different branch other than loris-integration-2. Is it? I tried it and I get errors

File: src/shared/consts.tsx (hunk 6 above)
Comment (ramirlm) [reply src/shared/consts.tsx]: Maybe the endpoints you mentioned as replaced, but I'll double check and understand why. If necessary I create a separate type

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/LorisIntegration.utils.ts (hunk 7 above)
Comment (ramirlm) [reply src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/LorisIntegration.utils.ts:56]: This is the way it's being saved in the state

File: src/modules/Dashboard/api/api.ts (hunk 1 above)
Comment (ramirlm) [reply src/modules/Dashboard/api/api.ts]: Maybe you're referencing a different backend branch, as I can see in the one I point to it is being used

Comment (ChaconC) [reply src/modules/Dashboard/api/api.ts]: The one that saves the integration is the one you added as `saveIntegrationToApi`, which takes type, config, etc as parameters. This is the dummy endpoint which didn't include any additional server information.

File: src/shared/consts.tsx (hunk 3 above)
Comment (ChaconC) [reply src/shared/consts.tsx:441]: yep

Comment (ChaconC) [reply src/shared/consts.tsx]: The section I mentioned to include as example is for the `configuration` param (currently declared as `SaveIntegrationParams`)

File: src/modules/Dashboard/api/api.ts (hunk 1 above)
Comment (ramirlm) [reply src/modules/Dashboard/api/api.ts]: Got it! Makes sense now, will change. Thanks!

File: src/shared/consts.tsx (hunk 6 above)
Comment (ramirlm) [reply src/shared/consts.tsx]: CleanShot 2024-10-14 at 14 58 18@2x
CleanShot 2024-10-14 at 14 59 28@2x

I'll create a different contstant for it for now, as you can see in the screenshot it only allows the UpperCase LORIS.

Comment (ChaconC) [reply src/shared/consts.tsx]: My question was more related to which one is using lowercase? I know the backend expects uppercase.

File: src/modules/Builder/api/api.types.ts (hunk 5 above)
Comment (ChaconC) [reply src/modules/Builder/api/api.types.ts:74]: Thanks, confirmed the GET request expects snake case.

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    renderWithProviders(<ConfigurationPopup {...defaultProps} />, {
+      preloadedState,
+    });
+
+    await userEvent.type(screen.getByLabelText(/LORIS Server Hostname/i), 'hostname');
+    await userEvent.type(screen.getByLabelText(/LORIS Username/i), 'jane.doe');
+    await userEvent.type(screen.getByLabelText(/LORIS Password/i), '123456');
+    await userEvent.click(screen.getByTestId('loris-configuration-popup-submit-button'));
+
+    // TODO FAILING
```
Comment (ChaconC) [inline src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx]: Is this TODO still valid?

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    await userEvent.type(screen.getByLabelText(/LORIS Server Hostname/i), 'hostname');
+    await userEvent.type(screen.getByLabelText(/LORIS Username/i), 'jane.doe');
+    await userEvent.type(screen.getByLabelText(/LORIS Password/i), '123456');
+    await userEvent.click(screen.getByTestId('loris-configuration-popup-submit-button'));
+
+    await waitFor(() => {
+      expect(screen.getByTestId('loris-project-select')).toBeInTheDocument();
+    });
+
+    await userEvent.selectOptions(screen.getByTestId('loris-project-select'), 'Project1'); // FAILING
```
Comment (ChaconC) [inline src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx]: Is this TODO still valid?

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+  const saveConfiguration = async () => {
+    try {
+      const { hostname, username, password } = getValues();
+      const projects = await fetchLorisProjects(hostname, username, password);
+      setProjects(projects);
+      setStep(ConfigurationsSteps.SelectProject);
+      setError(undefined);
+    } catch (error) {
+      setError(t('loris.errors.fetchProjectsFailed'));
```
Comment (ChaconC) [inline src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx:61]: This error message is different than figma

| Figma | Actual |
|-|-|
| CleanShot 2024-10-17 at 16 04 40 | image |

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+      if (data.result && data.result[0].message.includes('has previously been tied to applet')) {
+        setError(t('loris.errors.appletAlreadyTied'));
+
+        return;
+      }
+
+      // Set the form values
+      setValue('hostname', hostname);
+      setValue('username', username);
+      setValue('project', project);
```
Comment (ChaconC) [inline src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx]: why do you need to update the form after calling `saveLorisProject`?

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx (hunk 9 above)
Comment (ramirlm) [reply src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx]: Yes, this one is the one I mentioned not being able to make RTL triggering the dropdown =/

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx (hunk 8 above)
Comment (ramirlm) [reply src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.test.tsx]: No, I've removed it

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx (hunk 11 above)
Comment (ramirlm) [reply src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx]: good catch out! That's from the previous logic, removed.

File: src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx (hunk 10 above)
Comment (ramirlm) [reply src/modules/Builder/features/BuilderAppletSettings/IntegrationsListSetting/Integrations/LorisIntegration/ConfigurationPopup/ConfigurationPopup.tsx:61]: Fixed!

**Review by ChaconC — APPROVED** [review]
approved, the skipped tests will be address in M2-7669
